[#44] Use default configuration

qoelet opened the pull request.

ChShersh commented on src/Hintman/Config.hs:
```diff
@@ -44,9 +44,15 @@ hintmanConfiguationCodec = HintmanConfig
         pure defaultHintmanConfig
+
+    catchNoSuchFile :: IOException -> IO HintmanConfig
+    catchNoSuchFile _ =  do
```
To my taste: it would be nicer to use fancy `Exc` pattern-synonym from `relude` to handle multiple errors of different types:

* http://hackage.haskell.org/package/relude-0.4.0/docs/Relude-Exception.html#v:Exc

qoelet replied:
Noted, I'll check relude

vrom911 commented on src/Hintman/Config.hs:
```diff
@@ -44,9 +44,14 @@ hintmanConfiguationCodec = HintmanConfig
+            pure defaultHintmanConfig
+        _ -> do
+            print $ "Hintman failed to load config from " <> path
+            pure defaultHintmanConfig
```
I see the pattern in these two cases. Not to repeat yourself you probably can move the text to print into the function in the where clause and the body the will look like:

```haskell
print errorMessage
pure defaultHintmanConfig
```

or just
```haskell
defaultHintmanConfig <$ print errorMessage
```

qoelet replied:
:+1:

vrom911 commented on src/Hintman/Config.hs:
```diff
@@ -44,9 +44,14 @@ hintmanConfiguationCodec = HintmanConfig
+        Exc (_ :: IOException) -> do
+            print $ "Config '" <> path <> "' does not exist! Using default configuration"
+            pure defaultHintmanConfig
+        _ -> do
```
Also, we know that this is `TomlException`. For more pleasant experience working on the project, we probably want to provide the exception description in here, not just "Failed to load". What do you think?

qoelet replied:
Right, that makes sense. Do you have any suggestion on what the catch all clause here should be? Maybe a generic error message with a `show err`?

vrom911 replied:
Yes! I think this will work okay here 👍

qoelet replied:
OK, I'll work on an updated PR. Thanks @vrom911!

qoelet commented:
@vrom911 @ChShersh I reworked the PR, also added one more test case for `LoadTomlException`, let me know if this is better or if any other suggested changes are required.

vrom911 commented on src/Hintman/Config.hs:
```diff
@@ -44,9 +44,11 @@ hintmanConfiguationCodec = HintmanConfig
-        print $ "Hintman failed to load config from " <> path
-        pure defaultHintmanConfig
+    catchExcAndReturnDefaultConfig :: SomeException -> IO HintmanConfig
+    catchExcAndReturnDefaultConfig e = case e of
```
If you write like this

```suggestion
    catchExcAndReturnDefaultConfig e = defaultHintmanConfig <$ case e of
```

you won't need to make it in every case and the `<$` operator will still work okay.

qoelet replied:
Ooh :+1:

vrom911 commented on src/Hintman/Config.hs:
```diff
@@ -44,9 +44,11 @@ hintmanConfiguationCodec = HintmanConfig
 loadFileConfig path =
-    Toml.decodeFile hintmanConfiguationCodec path `catch` catchLoad
+    (Toml.decodeFile hintmanConfiguationCodec path `catch` catchExcAndReturnDefaultConfig)
+      `catch` catchExcAndReturnDefaultConfig
```
:thinking: Why it catches twice?

qoelet replied:
Fixed! Thanks!

ChShersh commented on src/Hintman/Config.hs:
```diff
@@ -44,9 +44,11 @@ hintmanConfiguationCodec = HintmanConfig
+    catchExcAndReturnDefaultConfig :: SomeException -> IO HintmanConfig
+    catchExcAndReturnDefaultConfig e = case e of
+        Exc (_ :: LoadTomlException) -> defaultHintmanConfig <$ print ("Encountered LoadTomlException while trying to load config from " <> path)
+        Exc (_ :: IOException) -> defaultHintmanConfig <$ print ("Config '" <> path <> "' does not exist! Using default configuration")
```
For outputting text to terminal it's better to use `putTextLn` function from `relude`, not `print`. `print` will add extra quotes

qoelet replied:
Noted, now the test assertions look nicer. But with `toText (show e)` I get a type checking error that tells me I need to add an annotation... slotting in `@String` work but is that the best way?

vrom911 replied:
@qoelet , `show` is polymorphic in `relude`. So when you use `putTextLn . show` it understands that the type is `Text` and you don't need to do anything else. But when you say `putTextLn . toText . show` he thinks so: Okay, for `putTextLn` I need `Text`; function `toText` gives me `Text` from polymorphic type `a`, and function `show` gives you this type `a` (because it is polimorphic) from the given `Exception` (or whatever). So how would it know what is this `a` should be? It could be anything.
Hope it makes the error clearer.

qoelet replied:
Yes it's clearer to me now, fixing the PR in a bit, thanks alot @vrom911, @ChShersh ... I'm definitely learning with this PR :)

ChShersh commented on test/Test/Config.hs:
```diff
@@ -0,0 +1,26 @@
+import System.IO.Temp (writeSystemTempFile)
+import Test.Hspec
+
+import Hintman.Config
```
All imports should be either qualified or have explicit import lists according to our style guide
